site/source/_posts/2019-12-09-web3-what-are-your-options.md

```diff
@@ -0,0 +1,194 @@
+
+For years, I harboured a deep hatred of JS, and actively whinged about it at every opportunity I got; being ***forced*** to use it in my daily work life.  Now however, I do have to say; over the last few years I have *softened* to JS, and I am much more comfortable in my own skin when having to use it.
+
+It goes without saying, the entire web is JS.  Look around you - JS.  View the source of this article - JS files.  Look at your own app's dependencies - JS.  
+
+JavaScript, specifically Node, really is in everything we use, and that now also applies to our wonderful world of Cryptocurrencies.
+
+As I mentioned briefly in my [***last*** article](https://embark.status.im/news/2019/11/28/nim-vs-crystal-part-3-cryto-dapps-p2p/), my ***next*** article series is going to be about building your first DApp – from start to finish.  Inevitably, the frontend of our DApp needs to be able to communicate with the Ethereum Network.  This is where [Web3.js](https://web3js.readthedocs.io/en/v1.2.4/index.html) comes into the mix.  `Web3.js` is a collection of APIs allowing us such functionality as:  Reading & Writing data from Smart Contracts, sending and receiving Ether, encrypting / decrypting wallets & data, and *a whole bunch* of other stuff too.  Basically, *most* of the backend functionality available on the Ethereum Network natively becomes available for use in the browser.
+
+
+This is how the `web3.js` library talks to the Ethereum Network:
+
+![Web3 JS Diagram](../assets/images/web3-js-diagram.png)
+*Image credit: [iotbl](https://iotbl.blogspot.com/2017/03/ethereum-and-blockchain-2.html)*
+
+So, now that the basics are covered, let's go over installing and using the `web3.js` library.
+
+
+# Installing Web3
+
+Installing `web3.js` is as simple as:
+
+```
+npm install web3
+```
+
+*One thing worth noting here*; is that (coming from an anti-js background), I kept getting a `cannot find web3 module` error when trying to import web3 into a Node console.  If you, like me, aren't a big js fan, this can be solved by first running the `npm init` command to ensure there is a `package.json` file in the cwd, and *then* you can run `npm install web3`, and it will work fine.  (I realise this is basic stuff – but actually for someone who's *tried* to avoid Node at all costs, it was initially confusing enough to have to Google.)
+
+I am working from a Mac here, but if you are working from Windows, the install process *can* be exactly the same, assuming you do have [Node & NPM installed](https://phoenixnap.com/kb/install-node-js-npm-on-windows).
+
+So, with `web3.js` installed, let's do some basic interactions with the Ethereum Network, and ***dive on in!***
+
+
+
+# Communicating with the Ethereum Network
+
+## Wallet Interaction
+
+For this article, we're going to use [Ganache](https://www.trufflesuite.com/ganache), for simplicity, as our local Blockchain.  By using Ganache, we can spin up a local Ethereum node, without having to write a single line of code!
```

Comment:
Actually, if Embark is already around, we can use `embark simulator` which is Ganache :)